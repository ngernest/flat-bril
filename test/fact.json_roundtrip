{
  "functions": [
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a"
          ],
          "dest": "x",
          "funcs": [
            "fact"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v13",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "funcs": [],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v8"
          ],
          "dest": "v9",
          "funcs": [
            "fact"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v5",
            "v9"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v10"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "fact",
      "type": "int"
    }
  ]
}
