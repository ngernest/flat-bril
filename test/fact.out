@main
	op: call 		dest: x 		type: int		args: ["a"]		funcs: ["fact"]
	op: print		args: ["x"]	
	op: const		dest: v13		type: int		value: 0	
@fact
	op: id   		dest: v1		type: int		args: ["a"]	
	op: const		dest: v2		type: int		value: 0	
	op: eq   		dest: v3		type: bool		args: ["v1", "v2"]	
	op: br   		args: ["v3"]		labels: ["then.0", "else.0"]
.then.0
	op: const		dest: v4		type: int		value: 1	
	op: ret  		args: ["v4"]	
.else.0
	op: id   		dest: v5		type: int		args: ["a"]	
	op: id   		dest: v6		type: int		args: ["a"]	
	op: const		dest: v7		type: int		value: 1	
	op: sub  		dest: v8		type: int		args: ["v6", "v7"]	
	op: call 		dest: v9		type: int		args: ["v8"]		funcs: ["fact"]
	op: mul  		dest: v10		type: int		args: ["v5", "v9"]	
	op: ret  		args: ["v10"]	
